File: 0235. Lowest Common Ancestor of a Binary Search Tree/Solution0235.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution0235:
    def lowestCommonAncestor0(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if p.val <= root.val <= q.val or q.val <= root.val <= p.val:
            return root
        if p.val < root.val:
            return self.lowestCommonAncestor(root.left, p, q)
        return self.lowestCommonAncestor(root.right, p, q)

    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        cur = root
        minVal = min(p.val, q.val)        
        maxVal = max(p.val, q.val)
        while cur:
            if cur.val < minVal:
                cur = cur.right
            elif cur.val > maxVal:
                cur = cur.left
            else:
                return cur
        return None

File: 0235. Lowest Common Ancestor of a Binary Search Tree/Solution0235.cpp

# include<iostream>
using namespace std;

 struct TreeNode {
     int val;
    TreeNode *left;
     TreeNode *right;
     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 };
 

class Solution {
public:
    TreeNode* lowestCommonAncestor0(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (root->val > p->val && root->val > q->val) {
            return lowestCommonAncestor(root->left, p, q);
        }
        if (root->val < p->val && root->val < q->val) {
            return lowestCommonAncestor(root->right, p, q);
        }
        return root;
    }
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        TreeNode* cur = root;
        while (cur != nullptr) {
            if (cur->val > p->val && cur->val > q->val) {
                cur = cur->left;
            } else if (cur->val < p->val && cur->val < q->val) {
                cur = cur->right;
            } else {
                return cur;
            }
        }
        return nullptr;
    }
};